src/android/AuthStateManager.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright 2017 The AppAuth for Android Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package edu.berkeley.eecs.emission.cordova.jwtauth;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.support.annotation.AnyThread;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.util.Log;
+
+import net.openid.appauth.AuthState;
+import net.openid.appauth.AuthorizationException;
+import net.openid.appauth.AuthorizationResponse;
+import net.openid.appauth.RegistrationResponse;
+import net.openid.appauth.TokenResponse;
+
+import org.json.JSONException;
+
+import java.lang.ref.WeakReference;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * An example persistence mechanism for an {@link AuthState} instance.
+ * This stores the instance in a shared preferences file, and provides thread-safe access and
+ * mutation.
+ */
+public class AuthStateManager {
+
+    private static final AtomicReference<WeakReference<AuthStateManager>> INSTANCE_REF =
+            new AtomicReference<WeakReference<AuthStateManager>>(new WeakReference<AuthStateManager>(null));
+
+    private static final String TAG = "AuthStateManager";
+
+    private static final String STORE_NAME = "AuthState";
+    private static final String KEY_STATE = "state";
+
+    private final SharedPreferences mPrefs;
+    private final ReentrantLock mPrefsLock;
+    private final AtomicReference<AuthState> mCurrentAuthState;
+
+    @AnyThread
+    public static AuthStateManager getInstance(@NonNull Context context) {
```

Comment:
did it give you any errors while creating a static variable that referenced a context? I got an error when I did, which is why the existing auth classes (e.g. `DummyDevAuth`) are singletons. Or is that fixed by the `WeakReference`?

You could just make it a constructor since you can restore the current state from the preferences. In the foreground, there will be one instance created when the plugin is initalized. And for the background, there are only a few calls at a time anyway...